op-acceptance-tests/tests/base/reth_geth_divergence_test.go

```diff
@@ -0,0 +1,204 @@
+package base
+
+import (
+	"context"
+	"fmt"
+	"math/big"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/devnet-sdk/system"
+	"github.com/ethereum-optimism/optimism/devnet-sdk/testing/systest"
+	"github.com/ethereum-optimism/optimism/op-service/testlog"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+)
+
+// TestRethGethDivergence checks that reth and geth do not diverge in terms of state root.
+// geth is considered the canonical/correct implementation.
+func TestRethGethDivergence(t *testing.T) {
+	systest.SystemTest(t, rethGethDivergenceTestScenario())
+}
+
+func rethGethDivergenceTestScenario() systest.SystemTestFunc {
+	const (
+		// testTimeout is the maximum time allowed for the test to run
+		testTimeout = 60 * time.Second
+
+		// blockCheckCount is the number of recent blocks to check for divergence
+		blockCheckCount = 5
+	)
+
+	return func(t systest.T, sys system.System) {
+		logger := testlog.Logger(t, log.LevelInfo)
+		logger.Info("Started reth-geth divergence test")
+
+		// Check all L2 chains
+		for i, chain := range sys.L2s() {
+			chainIndex := i
+			currentChain := chain
+			t.Run(fmt.Sprintf("Chain_%d", chainIndex), func(t systest.T) {
+				t.Parallel()
+				checkChainForDivergence(t, currentChain, logger.New("chain", chainIndex), testTimeout, blockCheckCount)
+			})
+		}
+	}
+}
+
+// checkChainForDivergence checks a single chain for reth/geth divergence
+func checkChainForDivergence(
+	t systest.T,
+	chain system.L2Chain,
+	logger log.Logger,
+	testTimeout time.Duration,
+	blockCheckCount uint64,
+) {
+	// Create context with timeout
+	ctx, cancel := context.WithTimeout(t.Context(), testTimeout)
+	defer cancel()
+
+	// Find geth and reth nodes
+	gethNode, rethNode, err := findGethAndRethNodes(chain, logger)
+	if err != nil {
+		t.Skip(err.Error())
+		return
+	}
+
+	logger.Info("Found both geth and reth nodes",
+		"geth", gethNode.Name(),
+		"reth", rethNode.Name())
+
+	// Get clients for both nodes
+	gethClient, rethClient, err := getNodeClients(gethNode, rethNode)
+	if err != nil {
+		t.Fatalf("Failed to get clients: %v", err)
+		return
+	}
+
+	// Determine the block range to check
+	startBlock, endBlock, err := getBlockRangeToCheck(ctx, gethClient, blockCheckCount)
+	if err != nil {
+		t.Fatalf("Failed to determine block range: %v", err)
+		return
+	}
+
+	// Check each block for divergence
+	for blockNumber := startBlock; blockNumber <= endBlock; blockNumber++ {
```

Comment:
Curious: why not just check for the divergence only at "endBlock" ( or `min(latestGethBlock, latestRethBlock)` )? 🤔

If any previous block diverges, it consequently ends up causing divergence for every next block.

And if a block isn't diverged, then that correlates to all of the previous blocks being un-diverged.

So, just checking the latestBlock / endBlock would tell if you divergence really ended up happening or not.

Reply:
Thanks Yash - no good reason, I just forgot about this heuristic. I'll make it use the "endBlock".